### PR TITLE
Update producer and consumer to allow sending and receiving Key/Value records

### PR DESCRIPTION
**Pull request by nicholastmosher**
Here's a list of changes in this PR:

- Added `Record::key(&self) -> &[u8]` and `Record::value(&self) -> &[u8]` in the consumer
- Removed `impl AsRef<[u8]> for Record` in consumer because we now want to explicitly distinguish between keys and values
- Added `TopicProducer::send_key_value_record(&self, key: K, value: V, partition: i32)` to producer
- Changed the representation of record keys in the protocol from `key: B` to `key: Option<B>` (where B is a byte buffer)
- Added a test that shows that decoding a record with an old encoding _just so happens_ to work as expected
  - Old records are always represented as "empty keys" in the form of zero-length buffers
  - Zero-length buffers are encoded as a length whose value is zero `0x00`.
  - The new record encoding represents "absent keys" in the form of `Option::None`
  - `Option::None` tag is encoded as `0x00` (as opposed to `Option::Some`, whose tag is `0x01`)
  - Therefore, decoding _old records with empty keys_ into _new records with absent keys_ works out-of-the-box
  - However, we must never attempt to decode a new record with any key using the older decoder.

Some potentially-open-questions:
- Are we happy with the new function names? These names are all potentially up for debate:
  - `Record::key`
  - `Record::value`
  - ~`TopicProducer::send_key_value_record`~
  - `TopicProducer::send`

### CODE REVIEW

**Comment by ajhunyady** [comment]
@nicholastmosher, is there a sample CLI on how to use this feature? We need to prepare a simple app or blog to introduce it.

**Comment by nicholastmosher** [comment]
@ajhunyady this PR does not yet add a CLI interface for using key/value records. It occurred to me that K/V on the CLI may take some significant additional design consideration. We can either hold this PR until the CLI is ready, or we can merge this and work on the CLI interface as a separate PR. The CLI should not require any additional changes to the Rust client, so it would be fine to merge this first and CLI second if we wanted to.

The primary unanswered question about the CLI is: How should a user distinguish between the contents of the keys and the values in their data? The reason this is tricky is because, until now, the CLI has largely not cared about the contents of messages - all records were sent as simple binary buffers, with the single exception to this being the ability to send records "line-by-line", interpreting newlines as a delimiter between records in a single file.

With Key/Value records, we now need a way for users to indicate what data should be treated as a Key and what data should be treated as a Value. I am thinking of opening an issue for tracking potential ideas about how to do this. Here are some initial thoughts of mine on this:

- One easy method would be to specify one file whose contents will be a key and one file whose contents will be a value:
  - E.g. `fluvio produce topic --key=key.txt --value=value.txt`
  - In its simplest form, this command invocation would send only a single record

- What if we wanted to specify multiple key/value records at the same time?
  - We might say that one file holds one-key-per-line and another file holds one-value-per-line
  - E.g. `fluvio produce topic --lines --key=key.txt --value=value.txt`
  - This relies on the assumption that they key/value data is textual, so would not work for binary data
  - This would rely on keys and values being written on corresponding lines in the respective files
  - Could we define some sort of globbing for key-files versus value-files? How would we associate the right keys with the right values?
    - `fluvio produce topic --keys="key-file-*.txt" --values="value-file-*.txt"`

- What if we wanted to co-locate key and value data in a single file?
  - We now need a way to encode keys and values separately in a single file
  - For textual data, something simple like CSV might be sufficient: keys and values are comma-separated, unique records are line-separated
  - We would need another solution for binary data

- Do we want to start adding support for specific types of encoded data?
  - We could add support for JSON/YAML/other object files
  - E.g. the entire file is treated as a record's value, and a keypath is given to "pick out" the data to use for the key
  - Perhaps `fluvio produce topic -f data.json --key-path=".name"`
  - We would need to choose some type of json-path DSL

There are probably more considerations as well. I will add these notes and anything else I can think of to an issue just for the CLI, but needless to say it is something that still needs more thinking-through.

**Comment by ajhunyady** [comment]
Hmmm... this looks very complicated to me and it shouldn't be.

**Comment by sehz** [comment]
For CLI, here is one way to do:
```
$ fluvio produce --key-seperator="="
name=John Doe 
country=US
```

**Comment by sehz** [comment]
We should simplify API much as possible.  
In most cases, K/V will be typical:
```
producer.send("name","doe")
```

For case, where only values is send:
```
producer.send(None,"doe")
```

**Comment by ajhunyady** [comment]
I suggest keeping the .send() for values and have an alternative API for send KV. That way the most common use case `send` will remain unchanged.

**Comment by sehz** [comment]
K/V will be more typical use case rather than other way around.

**Comment by ajhunyady** [comment]
In that case, send it is...

**Comment by ajhunyady** [comment]
> For CLI, here is one way to do:
> 
> ```
> $ fluvio produce --key-seperator="="
> name=John Doe 
> country=US
> ```

There should be some indication this is a key/value record.

**Comment by sehz** [comment]
`--kv` to indicate kv value.

```
$ fluvio produce --kv --separator "="
name=John Doe 
country=US
```

**Comment by nicholastmosher** [comment]
I have added basic CLI support for Key/Value to this PR. Here is the basic usage pattern:

```
fluvio-produce 0.5.1
Write messages to a topic/partition

By default, this reads input from stdin until EOF, and sends the contents as one
record. Alternatively, input files may be given with '-f' and input may be sent
line-by-line with '-l'. In any case, the producer continues reading until it
encounters EOF.

USAGE:
    fluvio produce [FLAGS] [OPTIONS] <topic>

FLAGS:
    -l, --lines    Send each line of input as its own record (using '\n')
    -q, --quiet    Do not print progress output when sending records
    -h, --help     Prints help information

OPTIONS:
    -p, --partition <integer>
            The ID of the Partition to produce to [default: 0]

        --key-separator <key-separator>
            Sends key/value records split on the first instance of the
            separator. Implies --lines
    -f, --files <files>...
            Paths to files to produce to the topic. If absent, producer will
            read stdin

ARGS:
    <topic>    The name of the Topic to produce to
```

And here are some examples of producing and consuming records with keys and values.

All examples are using a topic named `kv-example`.

### Producing from a File

```
$ cat data.txt
alice=Alice in Wonderland
bob=Bob the Builder
eve=Eve the Listener

$ cargo run --bin fluvio -- produce kv-example --key-separator "=" -f data.txt
data.txt:0 alice=Alice in Wonderland
data.txt:1 bob=Bob the Builder
data.txt:2 eve=Eve the Listener
Ok!
```

### Producing from stdin

```
$ cargo run --bin fluvio -- produce kv-example --key-separator "="
bunny=The Easter Bunny
stdin:0 bunny=The Easter Bunny
zeus=God of Lightning
stdin:1 zeus=God of Lightning
^D
Ok!
```

### Consuming

```
$ cargo run --bin fluvio -- consume kv-example -B --key-separator="::"
alice::Alice in Wonderland
bob::Bob the Builder
eve::Eve the Listener
bunny::The Easter Bunny
zeus::God of Lightning
```

File: src/client/src/consumer.rs
Comment (simlay) [inline src/client/src/consumer.rs:587]: This is a breaking change so update the client `Cargo.toml` to `0.6.0` as the node client uses `as_ref`.

File: examples/00-produce/src/main.rs
Comment (nicholastmosher) [reply examples/00-produce/src/main.rs:29]: Good point, I will add an instance of each 👍

File: src/client/src/consumer.rs
Comment (nicholastmosher) [reply src/client/src/consumer.rs:587]: Yes, you're right. I would prefer to go ahead and make this change and bump the version to 0.7.0 unless anybody objects

File: examples/01-produce-key-value/src/main.rs
Comment (sehz) [inline examples/01-produce-key-value/src/main.rs:24]: Can we collapse this into produce example? (with different binary). Otherwise, we are starting to create too many example crates.
Or alternatively, combine with original produce example

File: examples/02-consume/src/main.rs
Comment (sehz) [inline examples/02-consume/src/main.rs:51]: Why this need ref since value returns ref

File: src/client/src/producer.rs
Comment (sehz) [inline src/client/src/producer.rs:59]: This is not optimized.  But looks this was previous implementation.  let's create issue to optimize this (not creating unnecessary virtual connection every time)

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [inline src/dataplane-protocol/src/record.rs:344]: Why remove this?

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [inline src/dataplane-protocol/src/record.rs:449]: Should not use `DefaultAsyncBuffer`.   It should be something like:
`DefaultRecord::new(key, value)`

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [inline src/dataplane-protocol/src/record.rs:487]: See comments previously

**Review by sehz — CHANGES_REQUESTED** [review]

File: examples/02-consume/src/main.rs
Comment (nicholastmosher) [reply examples/02-consume/src/main.rs:51]: I think the general pattern I see in rust is the following naming convention:
- `value(&self)` the simple name takes self by immutable reference
- `value_mut(&mut self)` the `_mut` name takes self by mutable reference
- `into_value(self)` the `into_` name consumes self by ownership

so I think `value` is the right choice here

File: src/dataplane-protocol/src/record.rs
Comment (nicholastmosher) [reply src/dataplane-protocol/src/record.rs:344]: Previously, we were using `Record` almost as if it were a newtype wrapped around `value`, a single piece of data in the Record. Now that we are using both key and value, it does not make sense to me that we have all of these conversions that are heavily biased towards `value` and ignore `key`. Plus this was largely unused anyways.

File: examples/02-consume/src/main.rs
Comment (sehz) [reply examples/02-consume/src/main.rs:51]: Actually, it was not about name.   
```
String::from_utf8_lossy(&bytes);
```
Since value() is ref, shouldn't be:
```
String::from_utf8_lossy(bytes);
```

File: src/client/src/producer.rs
Comment (nicholastmosher) [reply src/client/src/producer.rs:59]: https://github.com/infinyon/fluvio/issues/840

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [reply src/dataplane-protocol/src/record.rs:344]: It still useful and this breaks contract with previous version unnecessary.

File: examples/01-produce-key-value/src/main.rs
Comment (nicholastmosher) [reply examples/01-produce-key-value/src/main.rs:24]: The main reason I did not combine the two is because the regular `produce` example is actually used in a test. So there wasn't really a great way to keep the examples small and clean while also satisfying the test and putting both together.

Comment (sehz) [reply examples/01-produce-key-value/src/main.rs:24]: fair enough.

File: src/dataplane-protocol/src/record.rs
Comment (nicholastmosher) [reply src/dataplane-protocol/src/record.rs:344]: I went ahead and put this back 👍

File: examples/02-consume/src/main.rs
Comment (nicholastmosher) [reply examples/02-consume/src/main.rs:51]: Oh I see, you're right, I will fix that

Comment (nicholastmosher) [reply examples/02-consume/src/main.rs]: Fixed

File: src/client/src/consumer.rs
Comment (sehz) [inline src/client/src/consumer.rs:592]: revert back this as well

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [inline src/dataplane-protocol/src/record.rs:356]: ? it was from Vec<u8>

File: src/dataplane-protocol/tests/file_fetch.rs
Comment (sehz) [inline src/dataplane-protocol/tests/file_fetch.rs:124]: `records[0].key().to_string()`

File: src/dataplane-protocol/tests/file_fetch.rs
Comment (sehz) [inline src/dataplane-protocol/tests/file_fetch.rs:126]: records[1].key().unwrap().to_string()

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by sehz** [comment]
@simlay Can you review this as well?

File: src/dataplane-protocol/src/record.rs
Comment (nicholastmosher) [reply src/dataplane-protocol/src/record.rs:356]: Oh there were a handful of them I think. Do you want them all back?

File: src/dataplane-protocol/tests/file_fetch.rs
Comment (nicholastmosher) [reply src/dataplane-protocol/tests/file_fetch.rs:126]: There is no `key()` method for `dataplane::record::Record`. Do you want me to add one?

Comment (sehz) [reply src/dataplane-protocol/tests/file_fetch.rs:126]: Yes.  we should add them

File: src/dataplane-protocol/src/record.rs
Comment (sehz) [reply src/dataplane-protocol/src/record.rs:356]: Yes, don't need to break old client unnecessary

File: src/client/src/consumer.rs
Comment (sehz) [inline src/client/src/consumer.rs:585]: Actually, key and value should move to `DefaultRecord`.  
Can use `Deref` to delegate methods

Comment (nicholastmosher) [reply src/client/src/consumer.rs:585]: Hm, I think using Deref for anything except smart-pointers is [considered an anti-pattern](https://github.com/rust-unofficial/patterns/blob/master/anti_patterns/deref.md), but I can make this method just forward to the `key()` method of `DefaultRecord` 👍

Comment (sehz) [reply src/client/src/consumer.rs]: Regarding `Deref`.  I disagree.  It's being used frequently and I see no problem here.
Alternative, then just implement `inner()` which will return DefaultRecord.  Caller just need to do extra hop

Creating delegation method is really anti-pattern in my opinion.

File: src/client/src/producer.rs
Comment (simlay) [inline src/client/src/producer.rs:39]: What's up with this docstring?

File: src/client/src/lib.rs
Comment (simlay) [inline src/client/src/lib.rs:108]: nit: "sends records or keys/values to the named topic" or something.

**Review by simlay — CHANGES_REQUESTED** [review]
Sehyo managed to cover most of my concerns about things. I had some minor nits.

Also what's the file `src/dataplane-protocol/tests/test_old_record_empty_key.bin`?

File: src/client/src/producer.rs
Comment (nicholastmosher) [reply src/client/src/producer.rs:39]: So this code block will be compiled as Rust code to ensure we don't break the code in the example. In the rust doc-string code-blocks, any line starting with a `#` will not be displayed on the docs page. In this example, the only thing the user will see on docs.rs is 

```
producer.send("Key", "Value").await?;
```

they will not see the rest of this code block, which is nice to reduce visual boilerplate and get right to the meat of the example.

**Comment by nicholastmosher** [comment]
@simlay that `.bin` file is what I used to verify that records in the new schema can be successfully decoded from old records that were encoded with the old schema. See https://github.com/infinyon/fluvio/pull/828/files#diff-e50536c96490b842b8f91fab3fca42353b0beca28ebe99ecd1eb80bcfc53c4b6R568-R606

Basically, I used the old encoder to encode a specific record into bytes, and saved those bytes into `test_old_record_empty_key.bin`. Then I wrote a test which attempts to decode a new-schema-record from those same bytes, and assert that the data landed where it was supposed to in the new record.

File: src/client/src/lib.rs
Comment (nicholastmosher) [reply src/client/src/lib.rs:108]: When I think "record" I consider that to include both the key and the value. This seems like the shortest way to describe it, saying "records or key/values" seems clumsy, and this is the one-line doc description.

Comment (simlay) [reply src/client/src/lib.rs:108]: :+1:

File: src/client/src/producer.rs
Comment (simlay) [reply src/client/src/producer.rs:39]: Cool.

**Review by simlay — APPROVED** [review]
I think I'm good with this.

**Review by sehz — APPROVED** [review]

**Comment by sehz** [comment]
Can you squash commits and produce clean commit logs?

**Pull request merged**
